qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py

```diff
@@ -0,0 +1,351 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The QuadraticHamiltonian class."""
+
+from __future__ import annotations
+
+from typing import Optional
+
+import numpy as np
+import scipy.linalg
+from qiskit.quantum_info.operators.mixins import TolerancesMixin
+from qiskit_nature.operators.second_quantization.fermionic_op import FermionicOp
+
+
+def _is_hermitian(mat: np.ndarray, rtol: float = 1e-5, atol: float = 1e-8) -> bool:
+    return np.allclose(mat, mat.T.conj(), rtol=rtol, atol=atol)
+
+
+def _is_antisymmetric(mat: np.ndarray, rtol: float = 1e-5, atol: float = 1e-8) -> bool:
+    return np.allclose(mat, -mat.T, rtol=rtol, atol=atol)
+
+
+class QuadraticHamiltonian(TolerancesMixin):
+    r"""A Hamiltonian that is quadratic in the fermionic ladder operators.
+
+    A quadratic Hamiltonian is an operator of the form
+
+    .. math::
+        \sum_{p, q} M_{pq} a^\dagger_p a_q
+        + \frac12 \sum_{p, q}
+            (\Delta_{pq} a^\dagger_p a^\dagger_q + \text{h.c.})
+        + \text{constant}
+
+    where :math:`M` is a Hermitian matrix and :math:`\Delta` is an antisymmetric matrix.
+    """
+
+    def __init__(
+        self,
+        hermitian_part: Optional[np.ndarray] = None,
+        antisymmetric_part: Optional[np.ndarray] = None,
+        constant: float = 0.0,
+        num_modes: Optional[int] = None,
+        validate: bool = True,
+        rtol: float = 1e-5,
+        atol: float = 1e-8,
+    ) -> None:
+        r"""Initialize a QuadraticHamiltonian.
+
+        Args:
+            hermitian_part: The matrix :math:`M` containing the coefficients of the terms
+                that conserve particle number.
+            antisymmetric_part: The matrix :math:`\Delta` containing the coefficients of
+                the terms that do not conserve particle number.
+            constant: An additive constant term.
+            num_modes: Number of fermionic modes. This should be consistent with hermitian_part
+                and antisymmetric_part if they are specified.
+            validate: Whether to validate the inputs.
```

Comment:
What is the rationale of giving a user the choice here (performance over hermitian and antisymmetric checks?)? I imagine things will go bad if the inputs do not conform to certain expectations, though I see these checks are only done when both matrices are specified. Things like the hermitian and anti-symmetric tests are not done on individual inputs. Nor is num_modes checked, if given, if only a single matrix is supplied.

Reply:
Yes, the user should be allowed to turn off validation for performance reasons.

>Things like the hermitian and anti-symmetric tests are not done on individual inputs. Nor is num_modes checked, if given, if only a single matrix is supplied.

This is incorrect. The checks are, in fact, performed in those situations, and I have included tests for that behavior.

Reply:
My bad, I read the first if, under validate, of the and of the two and somehow misread the other tests as being nested under that. I see now that they are at the same level,